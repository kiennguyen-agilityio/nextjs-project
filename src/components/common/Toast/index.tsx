import React, { useEffect } from 'react';

interface ToastProps {
  message: string;
  type: 'success' | 'error' | '';
  onClose: () => void;
}

const Toast: React.FC<ToastProps> = ({ message, type, onClose }) => {
  useEffect(() => {
    const timer = setTimeout(() => {
      onClose();
    }, 3000);
    return () => clearTimeout(timer);
  }, [onClose]);

  return (
    <div
      className={`fixed top-4 right-4 px-4 py-2 rounded shadow-lg ${type === 'success' ? 'bg-green-500' : 'bg-red-500'} text-white`}
    >
      {message}
    </div>
  );
};

export default Toast;
