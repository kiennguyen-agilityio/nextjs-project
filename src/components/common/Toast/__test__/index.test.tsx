import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { useRouter } from 'next/navigation';
import Toast from '@/components/common/Toast';

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));

describe('Toast component', () => {
  const mockPush = jest.fn();
  beforeEach(() => {
    (useRouter as jest.Mock).mockReturnValue({ push: mockPush });
  });

  test('renders Toast with success type and message', () => {
    render(<Toast type="success" message="Success message" />);
    expect(screen.getByText('Success message')).toBeInTheDocument();
    expect(screen.getByRole('alert')).toHaveClass('bg-green-500');
  });

  test('renders Toast with error type and message', () => {
    render(<Toast type="error" message="Error message" />);
    expect(screen.getByText('Error message')).toBeInTheDocument();
    expect(screen.getByRole('alert')).toHaveClass('bg-red-500');
  });

  test('hides Toast after duration', async () => {
    render(
      <Toast
        type="success"
        message="Success message"
        autoDismiss
        duration={500}
      />,
    );
    await waitFor(
      () => expect(screen.queryByText('Success message')).toBeNull(),
      { timeout: 600 },
    );
  });

  test('allows manual closing of Toast', () => {
    render(
      <Toast type="success" message="Success message" autoDismiss={false} />,
    );
    fireEvent.click(screen.getByLabelText('Close'));
    expect(screen.queryByText('Success message')).toBeNull();
  });

  test('does not show Toast if autoDismiss is false', async () => {
    render(
      <Toast type="success" message="Success message" autoDismiss={false} />,
    );
    await new Promise((resolve) => setTimeout(resolve, 1000)); // wait for a moment to ensure autoDismiss doesn't happen
    expect(screen.getByText('Success message')).toBeInTheDocument();
  });

  test('matches the snapshot', () => {
    const { asFragment } = render(
      <Toast type="success" message="Success message" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
