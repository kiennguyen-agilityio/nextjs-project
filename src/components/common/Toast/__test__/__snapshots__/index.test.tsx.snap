// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="fixed top-5 right-5 z-50 flex items-center bg-green-500 text-white text-sm font-bold px-4 py-3 rounded-lg"
    role="alert"
  >
    <span
      class="mr-2"
    >
      <svg
        class="w-4 h-4"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m4.5 12.75 6 6 9-13.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </span>
    <span>
      Success message
    </span>
    <button
      aria-label="Close"
      class="ml-4"
    >
      <svg
        class="fill-current h-6 w-6 text-white"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14.348 14.849a1.2 1.2 0 01-1.697 0L10 12.199l-2.651 2.65a1.2 1.2 0 01-1.697-1.697l2.651-2.651-2.65-2.651a1.2 1.2 0 011.697-1.697L10 8.803l2.651-2.65a1.2 1.2 0 011.697 1.697L11.399 10l2.65 2.651a1.2 1.2 0 010 1.698z"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
